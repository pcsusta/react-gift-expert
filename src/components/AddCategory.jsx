import { useState } from 'react';

export const AddCategory = ({ onNewCategory }) => {

  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => { // Aqui se desestructura el "event" para que abajo en vez de "event.target.value", directo se use el "target"
    setInputValue( target.value );
  };

  const onSubmit = ( event ) => {
    event.preventDefault();

    if( inputValue.trim().length <= 1 ){ return }; // Esto permitirá evaluar que el contenido del input sea mayor a 1 caracter, Si es menor, regresará un "return" que no hará nada y asi no se ejecutará la sig linea de codigo donde agregamos el contenido del input al array

    // setCategories( (categories) => [ inputValue, ...categories ] ); // En este caso, para poder traer aun las "categories" que contiene el "setCategories", primero se usaría un "callback" (categories) para despues poder utilizarlas, sino se perderían los valores que traia el "categories"
    onNewCategory( inputValue.trim() );
    setInputValue(''); // Aqui hacemos que al ejecutar el submit, al final se limpie en input
  };

  return (
    <form onSubmit={ onSubmit }>
      <input 
      type="text"
      placeholder="Buscar gifs"
      value={ inputValue }
      onChange={ onInputChange } // Aqui se está solo poniendo la referencia "onInputChange" en vez de todo el codigo largo que sería " (event) => onInputChange(event) "
    />
    </form>
  )
}
