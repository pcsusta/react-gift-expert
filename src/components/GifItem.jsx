export const GifItem = ({ title, url, id }) => { // Se desestructura para usar el objeto "image" que venÃ­a en las propiedades
  return (
    <div className="card">
      <img src={ url } alt={ title } />
      <p>{ title }</p>
    </div>
  )
}
