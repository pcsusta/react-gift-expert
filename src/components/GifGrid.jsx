import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }) => {
  // no es recomendable poner la funciónes aqui dentro porque se vuelve a ejecutar cada vez que se vuelve a renderizar el componente

  // Esto es un "custom hook". En todos los hooks se suele llamar primero escribiendo "use" seguido del nombre que tendrá
  const { images, isLoading } = useFetchGifs( category );

  return (
    <>
      <h3>{ category }</h3>
      {
        // Forma 1 para hacer un "if" Operador ternario
        //isLoading ? ( <h2>Cargando...</h2> ) : null 

        // Forma 2. Con un "if corto" asi: valor_a_evaluar && (x cosa)
        // El "if corto" funciona asi: 1. Primero es el valor a evaluar (si es true o false). 2. La acción a hacer si es "FALSE" (se pone un && sino habrá nada). 3. La acción a hacer si es "TRUE".
        isLoading && ( <h2>Cargando...</h2> )
      }
      
      {/* NOTA: En un .jsx no se puede usar la propiedad "class" en un html porque es una palabra reservada. Se usa ' className="x-clase-css" ' para agregar una clase CSS */}
      <div className="card-grid"> 
        {
          images.map( ( image ) => ( // se desestructura para en vez de tener "images.map( image => ( ... ) ", se hace de la otra forma
            <GifItem 
            key={ image.id }
            { ...image } /* Este codigo es el operador spread que se encarga de expandir el "image" de sus propiedades que contienen, en este caso contene las variables title, url y id. Sería como hacer esto: title={ image.title } url={ image.url } id={ image.id } */
            />
          ))
        }
        
      </div>   
    </>
  )
}
