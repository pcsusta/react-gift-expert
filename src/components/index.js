/* Esto se le llama archivo indice o archivo de barril.
Aqui haría la exportación de los componentes/archivos necesarios y que se encuentran en esta carpeta "components"
y de esta forma ya no tendría que importarlos a cada uno de manera individual en donde se necesite.

Es decir en vez de importarlos asi:
import { AddCategory } from "./components/AddCategory"; 
import { GifGrid } from "./components/GifGrid"; 

Gracias a éste archivo index.js, ahora podría resumirlos asi:
import { AddCategory, GifGrid } from "./components"; 
*/

export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';