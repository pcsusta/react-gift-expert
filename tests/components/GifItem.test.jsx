import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Pruebas en <GifItem.jsx />', () => { 

  const title = 'Madokami';
  const url   = 'http://madokami.co.js/madoka.jpg';

  test('Debe de hacer match con el Snapshot', () => {

    const { container } = render( <GifItem title={ title } url={ url } /> );
    expect( container ).toMatchSnapshot();

  });

  test('should debe de mostrar la imagen con el URL y el AT indicado', () => { 
  
    render( <GifItem title={ title } url={ url } /> );
    // screen.debug();

    // Esta es una forma pero se repite mucho codigo, en este caso se repite mucho el "screen.getByRole"
    // expect( screen.getByRole('img').src ).toBe( url );
    // expect( screen.getByRole('img').alt ).toBe( title );

    // En esta forma se utiliza una sola vez el "screen.getByRole" para usar un poco de menos codigo y es mas util por ejemplo en el caso de querer evaluar muchas mas propiedades.
    const { src, alt } = screen.getByRole('img');
    expect( src ).toBe( url );
    expect( alt ).toBe( title );

  });

  test('debe de mostrar el titulo en el componente', () => { 
    
    render( <GifItem title={ title } url={ url } /> );
    expect( screen.getByText( title ) ).toBeTruthy(); // En este caso el toBeTruthy evalúa que el texto del "title" exista en el componente evaluado. NO TOMA EN CUENTA las propiedades/atributos de los elementos. Por ejemplo, en la imagen está en "title" como atributo alt="" pero este no cuenta como un texto, sino como un atributo por lo que fallaría la prueba, pero como el "title" está en un <p></p> entonces la prueba pasa pues SÍ existe como texto

  });

});