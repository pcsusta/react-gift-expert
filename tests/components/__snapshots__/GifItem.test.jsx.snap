// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem.jsx /> Debe de hacer match con el Snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Madokami"
      src="http://madokami.co.js/madoka.jpg"
    />
    <p>
      Madokami
    </p>
  </div>
</div>
`;
